business/namespaces.go

```diff
@@ -129,12 +127,103 @@ func (in *NamespaceService) GetNamespaces(ctx context.Context) ([]models.Namespa
 		labelSelectorExcludeValue = excludeLabelList[1]
 	}
 
+	namespaces := []models.Namespace{}
+
+	if clientFactory == nil {
+		var err error
+		clientFactory, err = kubernetes.GetClientFactory()
+		if err != nil {
+			log.Errorf("Error getting client factory: %v", err.Error())
+			return nil, fmt.Errorf("Error getting client factory")
+		}
+	}
+
+	clusterNames := clientFactory.GetClusterNames()
+
+	wg := &sync.WaitGroup{}
+	type result struct {
+		cluster string
+		ns      []models.Namespace
+		err     error
+	}
+	resultsCh := make(chan result)
+
+	go func() {
+		for _, cluster := range clusterNames {
+			wg.Add(1)
+			go func(c string) {
+				defer wg.Done()
+				list, error := in.GetNamespacesByCluster(c)
+				if error != nil {
+					log.Errorf("Error fetching Namespaces per cluster %s: %s", c, error)
+					resultsCh <- result{cluster: c, ns: nil, err: error}
+				} else {
+					resultsCh <- result{cluster: c, ns: list, err: nil}
+				}
+			}(cluster)
+		}
+		wg.Wait()
+		close(resultsCh)
+	}()
+
+	// Combine namespace data
+	for result := range resultsCh {
+		if result.err != nil {
+			log.Errorf("Error fetching Namespaces per cluster %s: %s", result.cluster, result.err)
+			// TODO: Return error if one failure?
+			return nil, result.err
```

Comment:
This is a little tricky for multicluster. The current single cluster implementation returns an error if `GetNamespaces` fails but for multicluster we don't necessarily want the user to know if there's a cluster they can't access that is part of the mesh. Perhaps when we initially create the user clients in the client factory we then do some kind of access check and if that fails then we don't include that cluster in the map of user clients?

Reply:
+1, I'll add a new test.

Reply:
> So in case the error came from the local cluster it should be an error, it will keep the compatibility.

Returning an error to keep existing functionality for the home cluster (single cluster) sounds good to me. I don't think we want to make a bunch of exceptions for single cluster vs. multicluster but this seems like a good one to make to preserve the existing behavior for single cluster. We may need to make this configurable in the future but that can be added later.

> In case of MC, we assume the user always has to have permissions in the cluster, right?

This is true for single cluster but won't necessarily be true for multi cluster. The user could have no permissions for the "home cluster" but have access instead to the "remote cluster(s)".

It'd be good to document behavior like this for when multicluster is ready for use. It's probably too premature to have a multicluster section in the docs right now but I can create a separate MC issue for docs and make a note of this there.

Reply:
Now GetNamespace uses the cluster